packages/opal-server/opal_server/scopes/pullers.py

```diff
@@ -0,0 +1,128 @@
+import dataclasses
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import cast
+
+import pygit2
+from pygit2 import discover_repository, clone_repository, Repository, RemoteCallbacks, KeypairFromMemory, Username, \
+    UserPass
+
+from opal_common.schemas.scopes import Scope
+from opal_common.schemas.sources import SourceAuthData, SSHAuthData, GitHubTokenAuthData, GitPolicySource
+
+
+class SourcePuller(ABC):
+    def check(self) -> bool:
+        return False
+
+    @abstractmethod
+    def pull(self):
+        pass
+
+    def diff(self) -> (str, str):
+        return None, None
+
+
+@dataclasses.dataclass
+class InvalidScopeSourceType(Exception):
+    invalid_type: str
+
+
+@dataclasses.dataclass
+class GitError(Exception):
+    error: str
+
+
+class GitSourcePuller(SourcePuller):
+    def __init__(self, scope_id: str, base_dir: Path, source: GitPolicySource):
+        self.scope_id = scope_id
+        self.base_dir = base_dir
+        self.source = source
+
+    def check(self) -> bool:
+        """
+        returns True if new commits are available, false if up-to-date
+        """
+        if discover_repository(str(self._get_repo_path())) is None:
+            return True
+
+        repo = Repository(self._get_repo_path())
+        callbacks = self._get_callbacks()
+
+        repo.remotes['origin'].fetch(callbacks=callbacks)
+        latest_hash = repo.references['refs/remotes/origin/master'].target
+
+        return latest_hash != repo.head.target
+
+    def pull(self):
+        try:
+            self._pull()
+        except pygit2.GitError as e:
+            raise GitError(str(e))
+
+    def diff(self) -> (str, str):
+        repo = Repository(self._get_repo_path())
+        return repo.head.target, repo.references['refs/remotes/origin/master'].target
+
+    def _pull(self):
+        repo_path = self._get_repo_path()
+        callbacks = self._get_callbacks()
+
+        if discover_repository(str(repo_path)) is None:
+            clone_repository(self.source.url, str(self.base_dir/self.scope_id), callbacks=callbacks)
+        else:
+            repo = Repository(repo_path)
+            repo.remotes['origin'].fetch(callbacks=callbacks)
+            repo.checkout(repo.references['refs/remotes/origin/master'].resolve().name)
```

Comment:
the tracked branch name is not necessarily `master`, this should come from a config var